File: src/foundry/engines/planner.py
"""Delivery planning stage.

Turns a *ready* ticket plus verified context into a coding-agent-ready
:class:`DeliveryPlan`. Hard rule enforced here: ``agent_instructions`` is only
populated when the ticket is genuinely ready to build (acceptance criteria
present) AND a confident repository exists. Otherwise the plan is produced for
humans but carries no instructions an agent could act on.
"""

from __future__ import annotations

from typing import Protocol

from foundry.schemas.analysis import TicketAnalysis
from foundry.schemas.context import ContextBundle
from foundry.schemas.plan import DeliveryPlan, ImplementationStep, TestPlan
from foundry.schemas.risk import RiskAssessment
from foundry.schemas.ticket import RawTicket

# Default forbidden globs for the coding agent (also enforced by policy).
DEFAULT_FORBIDDEN_GLOBS = ["infra/**", "migrations/**", "**/.env*", "**/secrets/**"]

# The guardrail block and the PR-handoff closing are rendered by Foundry, never
# by a model: every planner (template or LLM) shares this exact text so an LLM
# planner can enrich the *plan* but can never relax a constraint.
CONSTRAINTS_BLOCK = """\
Constraints:
- Do not modify files matching: {forbidden}
- Do not add dependencies unless explicitly required.
- Do not perform database migrations.
- Do not change auth, payment, PII or infrastructure code.
- Stop and ask for human input if the change grows beyond the stated scope."""

CLOSING_BLOCK = """\
When you are done, open a draft PR whose description summarises what changed,
why, how it was tested, and any follow-ups."""

_INSTRUCTION_TEMPLATE = (
    """\
You are working on Linear issue {issue_key}: {title}.

Goal:
{goal}

Scope:
{scope}

Out of scope:
{out_of_scope}

Repository:
{repo}

Branch:
{branch}

Implementation plan:
{steps}

"""
    + CONSTRAINTS_BLOCK
    + "\n\n"
    + CLOSING_BLOCK
    + "\n"
)


class DeliveryPlanner(Protocol):
    def plan(
        self,
        ticket: RawTicket,
        analysis: TicketAnalysis,
        context: ContextBundle,
        risk: RiskAssessment,
    ) -> DeliveryPlan: ...


def branch_name_for(ticket: RawTicket) -> str:
    """Deterministic, sanitised branch name for a ticket."""
    slug = "".join(
        c if c.isalnum() else "-" for c in ticket.title.lower()
    ).strip("-")
    while "--" in slug:
        slug = slug.replace("--", "-")
    slug = slug[:40].strip("-")
    key = (ticket.issue_key or ticket.issue_id).lower()
    return f"foundry/{key}-{slug}" if slug else f"foundry/{key}"


class TemplatePlanner:
    """Reference planner that assembles a structured plan from upstream artifacts."""

    def plan(
        self,
        ticket: RawTicket,
        analysis: TicketAnalysis,
        context: ContextBundle,
        risk: RiskAssessment,
    ) -> DeliveryPlan:
        best_repo = context.best_repository
        affected = [best_repo.repo] if best_repo else []

        steps = [
            ImplementationStep(
                step=i,
                description=f"Satisfy acceptance criterion: {criterion}",
                expected_output="Code + tests covering this criterion.",
            )
            for i, criterion in enumerate(analysis.acceptance_criteria, start=1)
        ]

        test_plan = TestPlan(
            unit_tests=[f"Cover: {c}" for c in analysis.acceptance_criteria],
            manual_checks=context.test_commands,
        )

        open_questions = list(analysis.missing_information)

        plan = DeliveryPlan(
            goal=analysis.summary,
            scope=list(analysis.acceptance_criteria),
            out_of_scope=["Anything not listed in the acceptance criteria."],
            affected_repositories=affected,
            expected_files_or_areas=[],  # populated by richer context enrichers
            implementation_steps=steps,
            test_plan=test_plan,
            rollback_considerations=["Revert the PR; no data migration involved."],
            open_questions=open_questions,
            agent_instructions=None,
        )

        if self._can_instruct_agent(analysis, context):
            plan.agent_instructions = self._render_instructions(
                ticket, plan, best_repo.repo
            )
        return plan

    @staticmethod
    def _can_instruct_agent(analysis: TicketAnalysis, context: ContextBundle) -> bool:
        return analysis.is_ready_to_build and context.has_confident_repository()

    @staticmethod
    def _render_instructions(
        ticket: RawTicket, plan: DeliveryPlan, repo: str
    ) -> str:
        steps = "\n".join(
            f"{s.step}. {s.description}" for s in plan.implementation_steps
        )
        return _INSTRUCTION_TEMPLATE.format(
            issue_key=ticket.issue_key or ticket.issue_id,
            title=ticket.title,
            goal=plan.goal,
            scope="\n".join(f"- {s}" for s in plan.scope) or "- (none)",
            out_of_scope="\n".join(f"- {s}" for s in plan.out_of_scope),
            repo=repo,
            branch=branch_name_for(ticket),
            steps=steps or "(derive from acceptance criteria)",
            forbidden=", ".join(DEFAULT_FORBIDDEN_GLOBS),
        )

File: dotnet/src/Foundry/Engines/Planner.cs
// Delivery planning stage.
//
// Turns a *ready* ticket plus verified context into a coding-agent-ready
// DeliveryPlan. Hard rule enforced here: AgentInstructions is only populated
// when the ticket is genuinely ready to build (acceptance criteria present)
// AND a confident repository exists. Otherwise the plan is produced for humans
// but carries no instructions an agent could act on.

using System.Text;
using Foundry.Schemas;

namespace Foundry.Engines;

public interface IDeliveryPlanner
{
    DeliveryPlan Plan(RawTicket ticket, TicketAnalysis analysis, ContextBundle context, RiskAssessment risk);
}

public static class Planning
{
    /// <summary>Default forbidden globs for the coding agent (also enforced by policy).</summary>
    public static readonly IReadOnlyList<string> DefaultForbiddenGlobs = new[]
    {
        "infra/**", "migrations/**", "**/.env*", "**/secrets/**",
    };

    /// <summary>Deterministic, sanitised branch name for a ticket.</summary>
    public static string BranchNameFor(RawTicket ticket)
    {
        var slugBuilder = new StringBuilder();
        foreach (var c in ticket.Title.ToLowerInvariant())
        {
            slugBuilder.Append(char.IsLetterOrDigit(c) ? c : '-');
        }
        var slug = slugBuilder.ToString().Trim('-');
        while (slug.Contains("--"))
        {
            slug = slug.Replace("--", "-");
        }
        slug = slug[..Math.Min(40, slug.Length)].Trim('-');
        var key = (string.IsNullOrEmpty(ticket.IssueKey) ? ticket.IssueId : ticket.IssueKey)
            .ToLowerInvariant();
        return slug.Length > 0 ? $"foundry/{key}-{slug}" : $"foundry/{key}";
    }
}

/// <summary>Reference planner that assembles a structured plan from upstream artifacts.</summary>
public sealed class TemplatePlanner : IDeliveryPlanner
{
    private const string InstructionTemplate = """
You are working on Linear issue {issue_key}: {title}.

Goal:
{goal}

Scope:
{scope}

Out of scope:
{out_of_scope}

Repository:
{repo}

Branch:
{branch}

Implementation plan:
{steps}

Constraints:
- Do not modify files matching: {forbidden}
- Do not add dependencies unless explicitly required.
- Do not perform database migrations.
- Do not change auth, payment, PII or infrastructure code.
- Stop and ask for human input if the change grows beyond the stated scope.

When you are done, open a draft PR whose description summarises what changed,
why, how it was tested, and any follow-ups.
""";

    public DeliveryPlan Plan(
        RawTicket ticket, TicketAnalysis analysis, ContextBundle context, RiskAssessment risk)
    {
        var bestRepo = context.BestRepository;
        var affected = bestRepo is null ? new List<string>() : new List<string> { bestRepo.Repo };

        var steps = analysis.AcceptanceCriteria
            .Select((criterion, index) => new ImplementationStep
            {
                Step = index + 1,
                Description = $"Satisfy acceptance criterion: {criterion}",
                ExpectedOutput = "Code + tests covering this criterion.",
            })
            .ToList();

        var testPlan = new TestPlan
        {
            UnitTests = analysis.AcceptanceCriteria.Select(c => $"Cover: {c}").ToList(),
            ManualChecks = context.TestCommands.ToList(),
        };

        var plan = new DeliveryPlan
        {
            Goal = analysis.Summary,
            Scope = analysis.AcceptanceCriteria.ToList(),
            OutOfScope = new[] { "Anything not listed in the acceptance criteria." },
            AffectedRepositories = affected,
            ExpectedFilesOrAreas = Array.Empty<string>(), // populated by richer context enrichers
            ImplementationSteps = steps,
            TestPlan = testPlan,
            RollbackConsiderations = new[] { "Revert the PR; no data migration involved." },
            OpenQuestions = analysis.MissingInformation.ToList(),
            AgentInstructions = null,
        };

        if (CanInstructAgent(analysis, context))
        {
            plan = plan with
            {
                AgentInstructions = RenderInstructions(ticket, plan, bestRepo!.Repo),
            };
        }
        return plan;
    }

    private static bool CanInstructAgent(TicketAnalysis analysis, ContextBundle context) =>
        analysis.IsReadyToBuild && context.HasConfidentRepository();

    private static string RenderInstructions(RawTicket ticket, DeliveryPlan plan, string repo)
    {
        var steps = string.Join("\n", plan.ImplementationSteps.Select(s => $"{s.Step}. {s.Description}"));
        var scope = plan.Scope.Count > 0
            ? string.Join("\n", plan.Scope.Select(s => $"- {s}"))
            : "- (none)";
        var outOfScope = string.Join("\n", plan.OutOfScope.Select(s => $"- {s}"));
        return InstructionTemplate
            .Replace("{issue_key}", string.IsNullOrEmpty(ticket.IssueKey) ? ticket.IssueId : ticket.IssueKey)
            .Replace("{title}", ticket.Title)
            .Replace("{goal}", plan.Goal)
            .Replace("{scope}", scope)
            .Replace("{out_of_scope}", outOfScope)
            .Replace("{repo}", repo)
            .Replace("{branch}", Planning.BranchNameFor(ticket))
            .Replace("{steps}", steps.Length > 0 ? steps : "(derive from acceptance criteria)")
            .Replace("{forbidden}", string.Join(", ", Planning.DefaultForbiddenGlobs));
    }
}
